File: Estruturas de Dados e Complexidade de Algoritmos/Algoritmos40/Alg40.js
/*
♦ Algoritmo 40: Algoritmo de implementação da Estrutura de Dados Não Linear Dinâmica (não Estática) de Grafo: Floyd-Warshall.
*/


class FloydWarshall {
    constructor() {
        this.distancias = new Map();
        this.predecessores = new Map();
    }

    adicionarVertice(vertice) {
        if (!this.distancias.has(vertice)) {
            this.distancias.set(vertice, new Map());
            this.predecessores.set(vertice, new Map());
        }
    }

    adicionarAresta(origem, destino, peso) {
        this.adicionarVertice(origem);
        this.adicionarVertice(destino);

        this.distancias.get(origem).set(destino, peso);
        this.predecessores.get(origem).set(destino, origem);

        this.distancias.get(destino).set(origem, peso);
        this.predecessores.get(destino).set(origem, destino);
    }

    floydWarshall() {
        for (let k of this.distancias.keys()) {
            for (let i of this.distancias.keys()) {
                for (let j of this.distancias.keys()) {
                    if (this.distancias.get(i).has(k) && this.distancias.get(k).has(j)) {
                        let novoPeso = this.distancias.get(i).get(k) + this.distancias.get(k).get(j);
                        if (!this.distancias.get(i).has(j) || novoPeso < this.distancias.get(i).get(j)) {
                            this.distancias.get(i).set(j, novoPeso);
                            this.predecessores.get(i).set(j, this.predecessores.get(k).get(j));
                        }
                    }
                }
            }
        }
    }

    obterDistancias() {
        return this.distancias;
    }

    obterPredecessores() {
        return this.predecessores;
    }
}

const floydWarshall = new FloydWarshall();

floydWarshall.adicionarAresta('A', 'B', 10);
floydWarshall.adicionarAresta('A', 'C', 15);
floydWarshall.adicionarAresta('B', 'D', 12);
floydWarshall.adicionarAresta('B', 'F', 15);
floydWarshall.adicionarAresta('C', 'E', 10);
floydWarshall.adicionarAresta('D', 'E', 2);
floydWarshall.adicionarAresta('D', 'F', 1);
floydWarshall.adicionarAresta('F', 'E', 5);

floydWarshall.floydWarshall();

const distancias = floydWarshall.obterDistancias();
const predecessores = floydWarshall.obterPredecessores();

for (let [u, value_u] of distancias) {
    for (let [v, peso] of value_u) {
        console.log(`Distância de ${u} até ${v}: ${peso}`);
    }
}

for (let [u, value_u] of predecessores) {
    for (let [v, predecessor] of value_u) {
        console.log(`Predecessor de ${v} no caminho de ${u} até ${v}: ${predecessor}`);
    }
}

File: Estruturas de Dados e Complexidade de Algoritmos/Algoritmos40/Alg40.java
/*
♦ Algoritmo 40: Algoritmo de implementação da Estrutura de Dados Não Linear Dinâmica (não Estática) de Grafo: Floyd-Warshall.
*/


import java.util.HashMap;
import java.util.Map;

class FloydWarshall<T> {
    private Map<T, Map<T, Integer>> distancias;
    private Map<T, Map<T, T>> predecessores;

    public FloydWarshall() {
        distancias = new HashMap<>();
        predecessores = new HashMap<>();
    }

    public void adicionarVertice(T vertice) {
        if (!distancias.containsKey(vertice)) {
            distancias.put(vertice, new HashMap<>());
            predecessores.put(vertice, new HashMap<>());
        }
    }

    public void adicionarAresta(T origem, T destino, int peso) {
        adicionarVertice(origem);
        adicionarVertice(destino);

        distancias.get(origem).put(destino, peso);
        predecessores.get(origem).put(destino, origem);

        distancias.get(destino).put(origem, peso);
        predecessores.get(destino).put(origem, destino);
    }

    public void floydWarshall() {
        for (T k : distancias.keySet()) {
            for (T i : distancias.keySet()) {
                for (T j : distancias.keySet()) {
                    if (distancias.get(i).get(k) != null && distancias.get(k).get(j) != null) {
                        int novoPeso = distancias.get(i).get(k) + distancias.get(k).get(j);
                        if (distancias.get(i).get(j) == null || novoPeso < distancias.get(i).get(j)) {
                            distancias.get(i).put(j, novoPeso);
                            predecessores.get(i).put(j, predecessores.get(k).get(j));
                        }
                    }
                }
            }
        }
    }

    public Map<T, Map<T, Integer>> obterDistancias() {
        return distancias;
    }

    public Map<T, Map<T, T>> obterPredecessores() {
        return predecessores;
    }
}

public class Alg40 {
    public static void main(String[] args) {
        FloydWarshall<Character> floydWarshall = new FloydWarshall<>();

        floydWarshall.adicionarAresta('A', 'B', 10);
        floydWarshall.adicionarAresta('A', 'C', 15);
        floydWarshall.adicionarAresta('B', 'D', 12);
        floydWarshall.adicionarAresta('B', 'F', 15);
        floydWarshall.adicionarAresta('C', 'E', 10);
        floydWarshall.adicionarAresta('D', 'E', 2);
        floydWarshall.adicionarAresta('D', 'F', 1);
        floydWarshall.adicionarAresta('F', 'E', 5);

        floydWarshall.floydWarshall();

        Map<Character, Map<Character, Integer>> distancias = floydWarshall.obterDistancias();
        Map<Character, Map<Character, Character>> predecessores = floydWarshall.obterPredecessores();

        for (char u : distancias.keySet()) {
            for (char v : distancias.get(u).keySet()) {
                System.out.println("Distância de " + u + " até " + v + ": " + distancias.get(u).get(v));
            }
        }

        for (char u : predecessores.keySet()) {
            for (char v : predecessores.get(u).keySet()) {
                System.out.println("Predecessor de " + v + " no caminho de " + u + " até " + v + ": " + predecessores.get(u).get(v));
            }
        }
    }
}
